### DISCUSSION TITLE
Building Flutter App for Android throws an error: error: invalid source release: 17

### DISCUSSION THREAD

**Discussion by samnbd** [codemagic-ci-cd/codemagic-docs#2333] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1.Already have a Flutter app (using Flutter 3.7.11) that was building correctly before
2. Added [nordic_dfu](https://pub.dev/packages/nordic_dfu) Flutter package to my app (latest nordic_dfu version) 
3. Generate Android build on Codemagic

### Expected results

I expect to build the app correctly without issues.
FYI, the build is being generated correctly on my local machine, but I'm using Java 18.

### Actual results

The build fails with the following error: 

```
Execution failed for task ':nordic_dfu:compileReleaseJavaWithJavac'.
> error: invalid source release: 17
```
And I can't add Java 18 to the environment variables in Codemagic since it's not available as mentioned [here](https://docs.codemagic.io/specs/versions-linux/#java-versions)

**Comment by mikhail-tokarev** [comment]
Hi

sorry for the late reply. We will pre-install latest Java LTS on our machines, hopefully it will be available next week.

Thank you for bringing this to our attention and sorry again for the delay.

**Comment by icarusdust** [comment]
Hi @samnbd please note that Java 17 is available on Linux instances. You can specify it in` codemagi.yaml `like:

```
environment:
   vars: 
   java: 17
```

or in the Workflow Editor under the Environment variables section by defining an env var `JAVA_HOME` and setting its value to `/usr/lib/jvm/java-1.17.0-openjdk-amd64`

**Reply by goncalo-oliveira** [reply]
Setting this variable on the Workflow Editor doesn't really work. The machine initialization fails stating it's not a valid format. Maybe this is different for M1 machines?

**Reply by icarusdust** [reply]
The path I shared above is for setting Java with Linux instances. 

The path for Mac environment is different: `/Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home`

**Comment by mohammedbabelly20** [comment, marked as answer]
Codemagic machines now have **Java 21 (LTS)** installed. Check out the machines spec in the [docs](https://docs.codemagic.io/specs/versions-macos/).
